docs/spec_draft.md

```diff
@@ -3080,6 +3124,50 @@ More formally, for each element `x`: `real(x) = is_complex(x) ? x.real : x`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.recv
+
+### Semantics
+
+Receives data into a partition from a [stablehlo.send](#stablehlosend)
+instruction, in another partition, that shares the same channel handle pair
+{`channel_id`, `channel_type`}.
+
+If `is_host_transfer` is `true`, then the operation transfers data from the
+host.
+
+### Inputs
+
+| Name               | Type                                                               |
+|--------------------|--------------------------------------------------------------------|
+| `token`            | `token`                                                            |
+| `channel_id`       | constant of type `si64`                                            |
+| `channel_type`     | enum of `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, and `HOST_TO_DEVICE` |
+| `is_host_transfer` | constant of type `i1`                                              |
+
+### Outputs
+
+| Name      | Type                                                      |
+|-----------|-----------------------------------------------------------|
+| `results` | variadic number of tensors of any supported type or token |
+
+### Constraints
+  * (C1) [todo](https://github.com/openxla/stablehlo/issues/579) `channel_type` must be
+    * `HOST_TO_DEVICE`, if `is_host_transfer` $=$ `true`,
+    * `DEVICE_TO_DEVICE`, otherwise.
+  * (C2) size(`results`) > 0 if `is_host_transfer` $=$ `true`.
+  * (C3) type(`results`[-1]) $=$ `token`.
+
+### Examples
+
+```mlir
+%result:2 = "stablehlo.recv"(%token) {
+  channel_handle = #stablehlo.channel_handle< handle = 5, type = 3>,
+  is_host_transfer = true
+} : (!stablehlo.token) -> (tensor<3x4xi32>, !stablehlo.token)
+```
```

Comment:
Proposed rewording:

```
  %results:2 = "stablehlo.recv"(%token) {
    // channel_id = 5 : i64,
    // channel_type = #stablehlo<channel_type HOST_TO_DEVICE>,
    channel_handle = #stablehlo.channel_handle<handle = 5, type = 3>,
    is_host_transfer = true
  } : (!stablehlo.token) -> (tensor<3x4xi32>, !stablehlo.token)
```

Feedback:
  1. Typo: "results".
  2. Typo: extraneous whitespace between opening angle bracket and "handle".
  3. Added a comment that explains how `channel_handle` in the dialect decomposes into id and type.